src/server.ts

```diff
@@ -275,32 +282,43 @@ const sendDiagnostics = (ecosystem: string, uri: string, contents: string, colle
-            connection.sendDiagnostics({uri: uri, diagnostics: diagnostics});
         });
         
         let totalCount = new TotalCount();
 
-        for (let dependency of deps) {
-            if(dependency.name.value && dependency.version.value && regexVersion.test(dependency.version.value.trim())) {
-                get_metadata(ecosystem, dependency.name.value, dependency.version.value).then((response) => {
-                    if (response != null) {
-                        let pipeline = new DiagnosticsPipeline(DiagnosticsEngines, dependency, config, diagnostics, uri);
-                        pipeline.run(response);
-                        for (const item of pipeline.items) {
-                            let secEng = item as SecurityEngine;
-                            totalCount.vulnerabilityCount += secEng.vulnerabilityCount;
-                            totalCount.advisoryCount += secEng.advisoryCount;
-                        }
-                    }
-                    aggregator.aggregate(dependency);
-                }).catch((err)=>{
-                    aggregator.aggregate(dependency);
-                    connection.console.log(`Error ${err} while ${dependency.name.value}:${dependency.version.value}`);
+        constructPayload(deps, aggregator).then((requestPayload) => {
+            let payloadData: any;
+            payloadData = requestPayload;
+
+            for (let i = 0; i < payloadData.length; i += 10) {
+                let data = {
+                    "ecosystem": ecosystem,
+                    "package_versions": payloadData.slice(i, i + 10)
+                };
+                let reqData = JSON.stringify(data);
+                batchComponentAnalysis(reqData).then((response) => {
+                    let componentAnalysisResponse: any;
+                    componentAnalysisResponse = response;
+                    componentAnalysisResponse.filter((r) => {
+                        deps.filter((dep) => {
+                            if (r.package === dep.name.value && r.version === dep.version.value) {
+                                let pipeline = new DiagnosticsPipeline(DiagnosticsEngines, dep, config, diagnostics, uri);
+                                pipeline.run(r);
+                                for (const item of pipeline.items) {
+                                    let secEng = item as SecurityEngine;
+                                    totalCount.vulnerabilityCount += secEng.vulnerabilityCount;
+                                    totalCount.advisoryCount += secEng.advisoryCount;
+                                }
+                                connection.sendDiagnostics({ uri: uri, diagnostics: diagnostics });
+                                aggregator.aggregate(dep);
+                            }
+                        });
+                    });

```

Comment:
Instead of finding dependency object from request, you can construct one from the response and avoid O(n^2) looping. 
```suggestion
                    const dependency: IDependency = {
                      name: r.package,
                      version: r.version
                    };
                    const pipeline = new DiagnosticsPipeline(DiagnosticsEngines, dependency, config, diagnostics, uri);
                    pipeline.run(r);
                    for (const item of pipeline.items) {
                        const secEng = item as SecurityEngine;
                        totalCount.vulnerabilityCount += secEng.vulnerabilityCount;
                        totalCount.advisoryCount += secEng.advisoryCount;
                    }
                    connection.sendDiagnostics({ uri: uri, diagnostics: diagnostics });
                    aggregator.aggregate(dep);


```

Reply:
But here we need the position of the dependency version to set the range of diagnostic to flag the vulnerability with underline in the manifest file, and we can not get it from response.

here: https://github.com/fabric8-analytics/fabric8-analytics-lsp-server/blob/master/src/consumers.ts#L177

Reply:
Gotcha. BTW, why can't we can have a Map where key={package + version} and value is instance of IDependency. It would help you to get rid of O(n^2) iteration.

Reply:
I was thinking for the same, let me try and update the PR.